feat(local): add --no-watch flag to disable hot reload

roger-zhangg opened the pull request:
#### Which issue(s) does this change fix?

#8782

#### Why is this change necessary?

The AWS SAM CLI file watcher caused unstable behavior during local development with `--warm-containers`:

- **Multiple Lambda functions sharing a `CodeUri`** — a single change in any function falsely invalidated *all* of them, killing every running warm container and demoting back to LAZY mode. Reproed in LoopBack apps that auto-create temporary `.sandbox` directories (8+ minute startup times).
- **High-overhead file watching environments** — Microsoft Defender on Windows, large monorepos, etc. make the file observer dominate CPU/IO and effectively block local invocation. See #8782.

#### How does it address the issue?

Adds a `--no-watch` flag to `sam local start-api` and `sam local start-lambda`. When supplied with `--warm-containers`, SAM CLI skips creating any file observers (template watcher and the per-function `LambdaFunctionObserver`). Local code or template changes will not restart the running container — stop and rerun the command to pick up changes.

The flag is forwarded as a kwarg through `RefreshableSamFunctionProvider` and `WarmLambdaRuntime`. With `no_watch=True`, no observer is constructed, and every `watch` / `unwatch` / `start` / `stop` call becomes a guarded no-op.

Also adds debug logging around `LambdaFunctionObserver` lock acquire / resource processing to help triage future watcher issues.

#### What side effects does this change have?

- `--no-watch` only takes effect with `--warm-containers`. When supplied without warm containers, the flag is ignored and a `WARNING` is logged. The cold-mode `SamFunctionProvider` (which doesn't accept `no_watch`) never receives the flag — addresses the earlier review concern about an uncaught error in that path.
- New kwarg `no_watch: Optional[bool] = False` on `InvokeContext`, `RefreshableSamFunctionProvider`, and `WarmLambdaRuntime`. Defaults preserve existing behavior; no caller is required to update.
- Refreshes a stale `WarmLambdaRuntime.__init__` docstring (`warm_containers` → current parameters) flagged in earlier review.
- New `no_watch` entry in `schema/samcli.json` for both `local_start_api` and `local_start_lambda`.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Review the [generative AI contribution guidelines](https://github.com/aws/aws-sam-cli/blob/develop/CONTRIBUTING.md#ai-usage)
- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [x] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

roger-zhangg committed:
first debug version

vicheey commented on samcli/lib/utils/file_observer.py:
```diff
@@ -415,6 +424,17 @@ class SingletonFileObserver(metaclass=Singleton):
     """
 
+    # Ignore patterns for file watching to avoid infinite loops and unnecessary events
+    DEFAULT_IGNORE_PATTERNS = [
```
Should we provide a way for customer to provide customize ignore pattern? If none is provided, we can use the DEFAULT_IGNORE_PATTERNS.

roger-zhangg replied:
Yes this should be the idea when we ship it. I will use this version to debug with customer first

roger-zhangg committed:
add --no-watch flag

vicheey commented:
DO NOT MERGE.

roger-zhangg committed:
add no-watch flag

roger-zhangg committed:
Merge branch 'develop' of github.com:aws/aws-sam-cli into hot-swap-fix

roger-zhangg committed:
black lint fix

roger-zhangg commented:
This failing test seems unrelated, but will take a look

roger-zhangg commented:
also passed on my local
```
(.venv) ruojiazh@842f577b9ad2 aws-sam-cli % pytest tests/smoke/test_all_commands.py::TestAllCommands -n 4
======== 510 passed in 111.96s (0:01:51) ========
```

roger-zhangg commented on samcli/local/lambdafn/runtime.py:
```diff
@@ -405,13 +407,23 @@ def __init__(self, container_manager, image_builder, observer=None, mount_symlin
             Instance of the ContainerManager class that can run a local Docker container
         image_builder samcli.local.docker.lambda_image.LambdaImage
             Instance of the LambdaImage class that can create am image
-        warm_containers bool
-            Determines if the warm containers is enabled or not.
+        observer
+            Optional observer for file watching
+        mount_symlinks bool
+            Optional. True if symlinks should be mounted in the container
+        no_mem_limit bool
+            Optional. True if memory limit should be disabled
```
yes

roger-zhangg committed:
Merge branch 'develop' into hot-swap-fix

roger-zhangg committed:
Merge branch 'develop' into hot-swap-fix

roger-zhangg committed:
Merge branch 'develop' into hot-swap-fix

roger-zhangg committed:
feedback

roger-zhangg committed:
Merge branch 'develop' into hot-swap-fix

Resolve conflicts and address PR feedback:

- Use kwargs (`function_logical_ids`, `no_watch`) on `RefreshableSamFunctionProvider` instead of brittle positional `args.extend(...)` (addresses valerena's review concern about uncaught errors when --no-watch is used without --warm-containers).
- When --no-watch is supplied without --warm-containers, log a warning, ignore the flag, and never forward `no_watch` to the cold-mode `SamFunctionProvider` (which doesn't accept it).
- Combine the merge-conflicting changes in `WarmLambdaRuntime.create()`: keep develop's threading lock + container_dns refactor and gate observer.watch/start/unwatch behind `self._observer is not None` so --no-watch is safe.
- Update --no-watch help text to clarify the --warm-containers prerequisite and the use case from issue #8782 (file watching causes high CPU/IO with Microsoft Defender / large monorepos / .sandbox dirs).
- Regenerate schema/samcli.json.
- Add unit tests covering: provider kwarg routing for warm mode, warning + skip for cold mode, RefreshableSamFunctionProvider observer skipped with no_watch=True, and WarmLambdaRuntime no_watch lifecycle no-ops.

roger-zhangg committed:
chore: black formatting for runtime.py post-merge

Apply black formatting after merge of develop into hot-swap-fix.

roger-zhangg committed:
chore: post-review cleanups for --no-watch

- Add no_watch docstring entry to RefreshableSamFunctionProvider.
- Initialize self._observer = None up front in RefreshableSamFunctionProvider so the
  type annotation is unconditional and the no-watch branch is implied.
- Drop incidental whitespace-only hunks (blank line in invoke_context.__enter__,
  blank line in SingletonFileObserver.__init__) flagged during review.

roger-zhangg commented on samcli/commands/local/cli_common/options.py:
```diff
@@ -268,6 +268,16 @@ def warm_containers_common_options(f):
             multiple=False,
         ),
+        click.option(
+            "--no-watch",
```
Thanks for the careful read — but I don't think this bug reproduces. A couple of clarifications:

**`--no-watch` is registered in `warm_containers_common_options`, not `local_common_options`.**

```python
# samcli/commands/local/cli_common/options.py
def warm_containers_common_options(f):
    """
    Warm containers related CLI options shared by "local start-api" and "local start_lambda" commands
    """
    warm_containers_options = [
        click.option("--warm-containers", ...),
        click.option("--debug-function", ...),
        click.option(
            "--no-watch",
            is_flag=True,
            default=False,
            ...
        ),
    ]
```

That decorator is only applied to `sam local start-api` and `sam local start-lambda`. `sam local invoke` (`samcli/commands/local/invoke/cli.py`) only applies `@invoke_common_options` and `@local_common_options`, so Click never sees `--no-watch` for that command and the `cli()` signature does not need a `no_watch` parameter.

**Verified against the dev install (`samdev`):**

```
$ samdev local invoke --help | grep -i no-watch
(no output — flag is not registered)

$ samdev local invoke --help | tail -3
                                    message generated by AWS SAM CLI and
                                    display timestamps.
    -h, --help                      Show this message and exit.

$ samdev local start-api --help | grep -A1 no-watch
    --no-watch                      Disable file watching for hot reload.
                                    Only applies when --warm-containers is

$ samdev local start-lambda --help | grep -A1 no-watch
    --no-watch                      Disable file watching for hot reload.
                                    Only applies when --warm-containers is
```

No `TypeError`, `local invoke` works as before.

**Should we extend `--no-watch` to `sam local invoke` anyway?**

I'd argue no:

- `sam local invoke` does not support `--warm-containers` — there is no `warm_container_initialization_mode` argument in its `cli()` / `do_cli()`, and it never instantiates `RefreshableSamFunctionProvider` / `WarmLambdaRuntime` / `LambdaFunctionObserver`.
- Each `local invoke` runs in cold mode: spin up a fresh container, invoke once, tear down. There is no file watcher in that lifecycle for the flag to disable.
- Adding the flag there would either be a silent no-op (confusing) or require another "ignored, warning" path identical to the warm/cold validation we already added in `InvokeContext.__enter__`. That's surface-area noise without a real use case behind it.

roger-zhangg committed:
Merge branch 'develop' into hot-swap-fix

roger-zhangg committed:
Merge branch 'develop' into hot-swap-fix

roger-zhangg commented:
Merge after LE are all out

roger-zhangg committed:
Merge branch 'develop' into hot-swap-fix

roger-zhangg committed:
add LE to no-watch test
